language/move-vm/runtime/src/loader.rs

```diff
@@ -665,57 +665,71 @@ impl Loader {
         data_store: &mut impl DataStore,
         log_context: &impl LogContext,
     ) -> VMResult<()> {
-        self.verify_module(module, data_store, false, log_context)
+        self.verify_module(module, &BTreeMap::new(), data_store, false, log_context)
     }
 
     fn verify_module(
         &self,
         module: &CompiledModule,
+        bundle_verified: &BTreeMap<ModuleId, CompiledModule>,
         data_store: &mut impl DataStore,
         verify_no_missing_modules: bool,
         log_context: &impl LogContext,
     ) -> VMResult<()> {
         bytecode_verifier::verify_module(&module)?;
         self.check_natives(&module)?;
 
-        let deps = module.immediate_dependencies();
+        let (bundle_deps, cached_deps): (Vec<_>, Vec<_>) = module
+            .immediate_dependencies()
+            .into_iter()
+            .partition(|module_id| bundle_verified.contains_key(module_id));
         let loaded_imm_deps = if verify_no_missing_modules {
-            self.load_dependencies_verify_no_missing_dependencies(deps, data_store, log_context)?
+            self.load_dependencies_verify_no_missing_dependencies(
+                cached_deps,
+                data_store,
+                log_context,
+            )?
         } else {
-            self.load_dependencies_expect_no_missing_dependencies(deps, data_store, log_context)?
+            self.load_dependencies_expect_no_missing_dependencies(
+                cached_deps,
+                data_store,
+                log_context,
+            )?
         };
-        self.verify_module_dependencies(module, loaded_imm_deps)
+        let all_imm_deps = loaded_imm_deps.iter().map(|module| module.module()).chain(
+            bundle_deps
+                .iter()
+                .map(|module_id| bundle_verified.get(module_id).unwrap()),
```

Comment:
If it's possible, it would be nice to get rid of this `unwrap` by only creating only `cached_deps` at 682...

Reply:
The `unwrap()` is removed, together with changing the way `bundle_dep` and `cache_dep` are derived.